onnxruntime/core/framework/library_handles.cc

```diff
@@ -0,0 +1,43 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/framework/library_handles.h"
+#include "core/common/logging/logging.h"
+#include <utility>
+
+namespace onnxruntime {
+LibraryHandles::~LibraryHandles() noexcept {
+  UnloadLibraries();
+}
+
+LibraryHandles::LibraryHandles(LibraryHandles&& other) noexcept : libraries_(std::move(other.libraries_)) {}
+
+LibraryHandles& LibraryHandles::operator=(LibraryHandles&& other) noexcept {
+  if (this != &other) {
+    UnloadLibraries();
+
+    libraries_ = std::move(other.libraries_);
+  }
+
+  return *this;
+}
+
+void LibraryHandles::Add(PathString library_name, void* library_handle) {
+  libraries_.push_back(std::make_pair(std::move(library_name), library_handle));
```

Comment:
What if the library_name was already registered before?

Reply:
The OS-specific library loading utilities for Windows, Linux, and macOS all implement reference counting. Therefore, from the OS's perspective, loading the same library multiple times is fine as long as the loads() are balanced by the same number of unloads(), which this implementation ensures.

From ORT's perspective, loading the same custom ops library multiple times would register duplicate custom op domains, which would result in the following exception.

> error: C++ exception with description "Domain already set in registry"

I think returning that error for duplicate calls to `session_options.RegisterCustomOpsLibrary` is ok. I can add a test that expects this error for duplicate calls.

Adding logic to allow duplicate calls to `session_options.RegisterCustomOpsLibrary` without failure seems complicated. OS-specific search paths may mean that different "library names" can correspond to the same library, and I'm not sure that we can always expect the same library handle to be returned for the same library.